driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -624,4 +579,35 @@ public boolean hasError() {
             return t != null;
         }
     }
+
+    private class ResponseBuffersCallback implements SingleResultCallback<ResponseBuffers> {
+        @Override
+        public void onResult(final ResponseBuffers result, final Throwable t) {
+            if (t != null) {
+                processFailedRead(t);
+            } else {
+                SingleResultCallback<ResponseBuffers> callback = null;
+                readerLock.lock();
+                try {
+                    isReading = false;
+                    callback = readQueue.remove(result.getReplyHeader().getResponseTo());
+
+                    if (callback == null) {
```

Comment:
It's possible that the recipient of this message has not put itself on the read queue yet, so we have to store the message for them.